lib/core/test/unit/Cardano/Wallet/Network/LightSpec.hs

```diff
@@ -0,0 +1,305 @@
+            (xs,_) -> Just (reverse xs)
+    , getAddressTxs = \_ _ _ -> pure ()
+    }
+  where
+    toPoint = chainPointFromBlockHeader
+    toHeight = fromIntegral . fromEnum . blockHeight
+
+data ChainHistory = ChainHistory !MockChain ![(DeltaChain, MockChain)]
+    deriving Eq
+
+instance Show ChainHistory where
+    show (ChainHistory c0 deltas) = unwords $ L.intersperse "->"
+        [ show s
+        | b <- c0 : map snd deltas
+        , let Hash s = headerHash $ NE.head b
+        ]
+
+instance Arbitrary ChainHistory where
+    arbitrary = sized genChainHistory
+
+genChainHistory :: Int -> Gen ChainHistory
+genChainHistory n = go 1 chain0 (ChainHistory chain0 []) 
+  where
+    chain0 = genesisBlock :| []
+    go j c1 (ChainHistory c0 deltas)
+        | j >= n = pure $ ChainHistory c0 (reverse deltas)
+        | otherwise = do
+            (delta, c2) <- genDeltaChain j c1
+            ds <- splitForward delta
+            go (j+1) c2 $ ChainHistory c0 $ [ (d,c2) | d <- reverse ds ] ++ deltas
+
+    splitForward (Forward blocks tip) = do
+        blockss <- subsequences $ NE.toList blocks
+        pure $ Forward <$> blockss <*> pure tip
+    splitForward x = pure [x]
+
+-- | Randomly split a list into subsequences.
+subsequences :: [a] -> Gen [NonEmpty a]
+subsequences [] = pure []
+subsequences xs = do
+    n <- choose (1,length xs)
+    (NE.fromList (take n xs) :) <$> subsequences (drop n xs)
+
+
+data DeltaChain
+    = Forward (NonEmpty Block) BlockHeader
+    | Backward BlockHeader
+    | Idle
+    deriving (Eq,Show)
+
+genesisBlock :: Block
+genesisBlock = BlockHeader
+    { slotNo = 0
+    , blockHeight = Quantity 0
+    , headerHash = mockHash 0
+    , parentHeaderHash = Nothing
+    }
+
+mockHash :: Int -> Hash "BlockHeader"
+mockHash = Hash . B8.pack . show
+
+mockStabilityWindow :: Int
+mockStabilityWindow = 5
+
+genDeltaChain :: Int -> MockChain -> Gen (DeltaChain, MockChain)
+genDeltaChain n chain1 = frequency [(2,forward), (1,backward)]
+  where
+    tip1 = NE.head chain1
+    forward = do
+        let newblock = BlockHeader
+                { slotNo = succ (slotNo tip1)
+                , blockHeight = succ (blockHeight tip1)
+                , headerHash = mockHash n
+                , parentHeaderHash = Just $ headerHash tip1
+                }
+            chain2 = newblock NE.<| chain1
+        pure (Forward (newblock :| []) (NE.head chain2), chain2)
+    backward = do
+        m <- choose (0, min (NE.length chain1 - 1) mockStabilityWindow)
+        let chain2 = NE.fromList (NE.drop m chain1)
+        pure (Backward (NE.head chain2), chain2)
+
+{-------------------------------------------------------------------------------
+    Monad that mocks time evolution of a MockChain.
+    Also keeps track of the state of potential chain followers.
+-------------------------------------------------------------------------------}
+data Free f a
+    = Free (f (Free f a))
+    | Pure a
```

Comment:
Why invent free monad? 

https://hackage.haskell.org/package/free-5.1.7/docs/Control-Monad-Free.html#t:Free

Reply:
Many reasons, actually. 😅 

* Main reason: The wallet does not yet import use a `free` package or one of its many variants. I didn't want to introduce a package-level idiom without team discussion, but I felt comfortable introduce a local idiom. (My preference there would actually be my own `operational`. 😅 )
* We already have another hand-baked monad, the [`LSQ` monad][1].
* The definition of `Free` is ~20, fairly simple lines. This is hard to beat.

  [1]: https://github.com/input-output-hk/cardano-wallet/blob/b24c7e855dae25b56adf786f93137f389db0ebee/lib/core/src/Ouroboros/Network/Client/Wallet.hs#L594-L614